keps/sig-auth/2579-psp-replacement/README.md

```diff
@@ -0,0 +1,1001 @@
+
+[custom security contexts]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/277-ephemeral-containers#configurable-security-policy
+
+One way this could be handled under the current model is:
+1. Exempt a special username (not one that can be authenticated directly) from policy enforcement,
+   e.g. `ops:privileged-debugger`
+2. Grant the special user permission to ONLY operate on the ephemeral containers subresource (it is
+   critical that they cannot create or update pods directly).
+3. Grant (real) users that should have privileged debug capability the ability to impersonate the
+   exempt user.
+
+We could consider ways to streamline the user experience of this, for instance adding a special RBAC
+binding that exempts users when operating on the ephemeral containers subresource (e.g. an
+`escalate-privilege` verb on the ephemeral containers subresource).
+
+<<[/UNRESOLVED]>>
+
+#### Other Pod Subresources
+
+Aside from ephemeral containers, the policy is not checked for any other Pod subresources (status,
+bind, logs, exec, attach, port-forward).
+
+Although annotations can be updated through the status subresource, the apparmor annotations are
+immutable and the seccomp annotations are deprecated and slated for removal in v1.23.
+
+### Pod Security Standards
+
+Policy level definitions are hardcoded and unconfigurable out of the box. However, the [Pod Security
+Standards] leave open ended guidance on a few items, so we must make a static decision on how to
+handle these elements:
+
+**HostPorts** - (baseline) HostPorts will be forbidden. This is a more niche feature, and violates
+the container-host boundary.
+
+**AppArmor** - (baseline) Allow anything except `unconfined`. Custom AppArmor profiles must be
+installed by the cluster admin, and AppArmor fails closed when a profile is not present.
+
+<<[UNRESOLVED]>>
+**SELinux** - (baseline)
+- (Peter Hunt) we could restrict to certain Type values i.e. ban `spc_t` (privileged) or only allow
+  `container_t` (only allowed to touch `container_file_t`, or container contents). That seems
+  consistent with how AppArmor and Seccomp work
+- (Jordan Liggitt) Remember that this admission plugin can't tell the difference between what the
+  user specified and what an earlier mutating admission plugin specified. Disallowing setting
+  selinux labels in the pod spec would also prevent another admission plugin from setting
+  constrained-but-shared-within-the-same-namespace selinux labels on the pod.
+- (Tim Allclair) It seems like we have 3 options here, none of them good:
+    1. Forbid setting any SELinux options (i.e. use the default) under the baseline (or restricted)
+       profiles. This probably breaks shared volumes.
+    2. Allow all SELinux options, and require a 3rd party controller to enforce custom SELinux
+       policy.
+    3. Add a configuration knob to allowlist (or deny) specific SELinux configurations.
```

Comment:
I think this option would be preferable for the baseline policy.

Reply:
field-level configuration is an explicit non-goal for this proposal, and makes the meaning of "baseline" inconsistent cluster-to-cluster

Reply:
Can the baseline policy include a hard-coded list of valid container selinux labels?

Reply:
This list could potentially change over time if container-selinux policy adds more types but could remain the same for a particular release.

Reply:
I spoke with @mrunalp, and it sounds like hardcoding an allowlist in a vendor-agnostic way might be feasible, and would definitely be preferable to the options enumerated here.

Reply:
That sounds ideal

Reply:
> I spoke with @mrunalp, and it sounds like hardcoding an allowlist in a vendor-agnostic way might be feasible, and would definitely be preferable to the options enumerated here.

any details to add here?

Reply:
@tallclair @liggitt 

These are the types that could be allowed at the baseline level to be configurable by the user.
```
 container_init_t
 container_kvm_t
 container_logreader_t
 container_t
```

Reply:
So to make sure I understand correctly, the following are allowed under the baseline policy:

- Unset SELinuxOptions (use the runtime defaults)
- Any value of `user`, `role` and/or `level`, as long as `type` is one of the values @mrunalp listed

Did I understand that correctly?

Reply:
> Unset SELinuxOptions (use the runtime defaults)

Yes :+1:

> Any value of user, role and/or level, as long as type is one of the values @mrunalp listed

We should allow changing the `level` (restricted to the list) and/or the `type` but not the `user` or the `role` for baseline policy.

Reply:
Does this look right?

> - Unset `SELinuxOptions` is allowed.
> - If options are set, they must match the following:
>   - `type` is one of:
>     - `container_init_t`
>     - `container_kvm_t`
>     - `container_logreader_t`
>     - `container_t`
>   - `level` can be anything
>   - `user` is unset
>   - `role` is unset

Reply:
Yes, that looks right :+1: